### PR TITLE
Rewrite of websocket code in TypeScript

### PR DESCRIPTION
**Pull request by kpalway**
Couple of things still TODO:
- Add cookie saving/loading code
- Merge with Larry's storage change
- Flesh out some of the placeholder types used (probably goes along with rewriting things like project-service.js)

### CODE REVIEW

File: src/frontend/src/helpers/websocket_client.ts [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+
+  private async authenticate(success: boolean, server_challenge: AuthResult): Promise<boolean> {
+    // Generate a nonce
+    var client_nonce = new Uint8Array(sjcl.random.randomWords(32));
+    for(var i=0; i<client_nonce.length; i++) {
+      client_nonce[i] = client_nonce[i] & 0xFF;
+    }
+
+    // OK, now we proceed to authenticate.
+    var raw_response = new Uint8Array(client_nonce.byteLength + server_challenge.byteLength);
```
Comment (e45lee) [inline src/frontend/src/helpers/websocket_client.ts]: Look at helpers/crypto.ts for the new authentication interface.

File: src/frontend/src/helpers/websocket_client.ts [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+    });
+  }
+
+  public close(): void {
+    this.websocket.close();
+  }
+
+  private async authenticate(success: boolean, server_challenge: AuthResult): Promise<boolean> {
+    // Generate a nonce
+    var client_nonce = new Uint8Array(sjcl.random.randomWords(32));
```
Comment (e45lee) [inline src/frontend/src/helpers/websocket_client.ts]: window.getRandomBytes (so we don't have to depend on another JavaScript library)

File: src/frontend/src/helpers/websocket_client.ts [hunk 3]
```diff
@@ -0,0 +1,469 @@
+import {Coder} from './crypto';
+import * as sjcl from 'sjcl';
+import {History,Settings,Change} from './types';
+
+class Message {
```
Comment (e45lee) [inline src/frontend/src/helpers/websocket_client.ts]: Easier thing to do here is to use tagged unions -- https://www.typescriptlang.org/docs/handbook/advanced-types.html

File: src/frontend/src/helpers/websocket_client.ts [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+
+    this.ready = new Promise((resolve, reject)=>{
+      this.websocket = new WebSocket(this.uri);
+      this.started = true;
+      this.websocket.onerror = ()=>{
+        this.failed = true;
+        if(!this.authenticated) {
+          reject("Socket closed during authentication!");
+        }
+        return this.closes && this.closes();
```
Comment (e45lee) [inline src/frontend/src/helpers/websocket_client.ts]: Errors should be propagated to all waiting promises.

File: src/frontend/src/helpers/websocket_client.ts [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+
+  /** Sends a message along the connection, ensuring that
+   *  the server and client are properly authenticated.
+   *
+   *  If the socket has not been properly authenticated,
+   *  sends the message after the socket has been properly
+   *  authenticated/set up. */
+  private async sendMessage(message: Message): Promise<WebsocketResult> {
+    if(this.failed || this.closed || !this.started) {
+      throw "Socket closed or failed!";
```
Comment (e45lee) [inline src/frontend/src/helpers/websocket_client.ts]: Exception class?

File: src/frontend/src/helpers/websocket.ts [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+      }
+      if(this.offlineEnabled()) {
+        var init = await this.storage.isInitialized();
+        if(!init && !this.isOffline()) {
+          await this.syncAll();
+        } else if(!init) {
+          throw new WebsocketError("Cannot call "+name+" in offline mode before an initial sync.");
+        }
+        var res = await this.storage[name].apply(this.storage, args);
+      }
```
Comment (YLiLarry) [inline src/frontend/src/helpers/websocket.ts]: Can anyone tell me why this is needed?
```
+      if(this.offlineEnabled()) {
 +        var init = await this.storage.isInitialized();
 +        if(!init && !this.isOffline()) {
 +          await this.syncAll();
 +        } else if(!init) {
 +          throw new WebsocketError("Cannot call "+name+" in offline mode before an initial sync.");
 +        }
 +        var res = await this.storage[name].apply(this.storage, args);
 +      }
```
Does this mean if the offline mode is on "enabled", and storage is dirty, `this.syncAll()` is never going to be called here for any of those `offline_enabled` functions?

Comment (e45lee) [inline src/frontend/src/helpers/websocket.ts]: It wouldn't make sense (say reading a file offline when you haven't synced...)

**Comment by e45lee** [comment]
See #687 for this PR.

**Pull request closed without merging**